lib/Runtime/Math/JavascriptMath.cpp

```diff
@@ -1016,68 +1016,86 @@ namespace Js
+                rand_s(reinterpret_cast<unsigned int*>(&s0.HighPart));
+                rand_s(reinterpret_cast<unsigned int*>(&s1.LowPart));
+                rand_s(reinterpret_cast<unsigned int*>(&s1.HighPart));
+
+                seed0 = s0.QuadPart;
+                seed1 = s1.QuadPart;
+            }
+        }
+
+        double ConvertRandomSeedsToDouble(const uint64 seed0, const uint64 seed1)
+        {
+            static const uint64 mExp  = 0x3FF0000000000000;
+            static const uint64 mMant = 0x000FFFFFFFFFFFFF;
+
+            // Take lower 52 bits of the sum of two seeds to make a double
+            // Subtract 1.0 to negate the implicit integer bit of 1. Final range: [0.0, 1.0)
+            // See IEEE754 Double-precision floating-point format for details
+            //   https://en.wikipedia.org/wiki/Double-precision_floating-point_format
+            uint64 resplusone_ui64 = ((seed0 + seed1) & mMant) | mExp;
+            double res = *(reinterpret_cast<double*>(&resplusone_ui64)) - 1.0;
+            return res;
+        }
+
+        void Xorshift128plus(uint64 *seed0, uint64 *seed1)
+        {
+            uint64 s1 = *seed0;
+            uint64 s0 = *seed1;
+            *seed0 = s0;
+            s1 ^= s1 << 23;
+            s1 ^= s1 >> 17;
+            s1 ^= s0;
+            s1 ^= s0 >> 26;
+            *seed1 = s1;
+        }
+
+        double JavascriptMath::Random(ScriptContext *scriptContext)
+        {
+            uint64 seed0 = scriptContext->GetLibrary()->GetRandSeed0();
+            uint64 seed1 = scriptContext->GetLibrary()->GetRandSeed1();
+
+            if (seed0 == 0 && seed1 == 0)
+            {
+                InitializeRandomSeeds(seed0, seed1);
 #if DBG_DUMP
                 if (Configuration::Global.flags.Trace.IsEnabled(PRNGPhase))
                 {
-                    Output::Print(L"[PRNG:%x] INIT %I64x\n", scriptContext, s0.QuadPart);
+                    Output::Print(L"[PRNG:%x] INIT %I64x %I64x\n", scriptContext, seed0, seed1);
                 }
 #endif
-
-                temp = s0.LowPart ^ multiplier.LowPart;
-                // Put bytes in order 0213.
-                temp = ((temp & 0xFF000000) >>24) | ((temp & 0x000000FF) <<24) | (temp & 0x00FFFF00);
-
-                // Interleave the bits : generator is 3120.
-                temp = ((temp & 0x0F000F00) >> 4) | ((temp & 0x00F000F0) << 4) | (temp & 0xF00FF00F);
-                temp = ((temp & 0x30303030) >> 2) | ((temp & 0x0C0C0C0C) << 2) | (temp & 0xC3C3C3C3);
-                temp = ((temp & 0x44444444) >> 1) | ((temp & 0x22222222) << 1) | (temp & 0x99999999);
-
-                s1.HighPart = temp >> 16;
-                s1.LowPart = (temp << 16) | ((s0.HighPart ^ s0.LowPart)& 0x0000FFFF);
-                seed = s1.QuadPart;
-
-                ThreadContext *threadContext = scriptContext->GetThreadContext();
-                threadContext->GetEntropy().AddThreadCycleTime();
```

Comment:
Given that rand_s() calls may fail, entropy class call is added back in as a backup.